lib/trino-parquet/src/test/java/io/trino/parquet/crypto/TestParquetEncryption.java

```diff
@@ -0,0 +1,611 @@
+        List<Integer> values = readSingleColumnFile(
+                file,
+                new TestingKeyRetriever(Optional.of(KEY_FOOT), Optional.empty(), Optional.empty()),
+                /*checkFooterIntegrity*/ true,
+                Optional.empty());
+
+        verifySequence(values, 100);   // data round‑trip OK
+    }
+
+    @ParameterizedTest(name = "aadPrefix_present={0}")
+    @CsvSource({"true", "false"})
+    void aadPrefixRoundTrip(boolean supplyPrefix)
+            throws IOException
+    {
+        File file = Files.createTempFile("pme‑aad", ".parquet").toFile();
+        file.deleteOnExit();
+
+        byte[] prefix = "fileAAD‑prefix".getBytes(UTF_8);
+        writeSingleColumnFile(file, /*compressed*/ true, /*encryptFooter*/ true, /*columnEncryptedWithFooterKey*/ false, Optional.of(prefix), () -> range(0, 100));
+
+        // build FileDecryptionProperties with / without the required prefix
+        FileDecryptionProperties.Builder properties = FileDecryptionProperties.builder()
+                .withKeyRetriever(new TestingKeyRetriever(Optional.of(KEY_FOOT), Optional.of(KEY_AGE), Optional.empty()));
+
+        if (supplyPrefix) {
+            properties.withAadPrefix(prefix);
+            // correct prefix => read succeeds
+            List<Integer> values = readSingleColumnFile(
+                    file,
+                    new TestingKeyRetriever(Optional.of(KEY_FOOT), Optional.of(KEY_AGE), Optional.empty()),
+                    true,                                       // footer integrity
+                    Optional.of(prefix));                       // pass prefix
+
+            verifySequence(values, 100);
+        }
+        else {
+            // missing / wrong prefix => should fail while reading footer
+            assertThatThrownBy(() -> readSingleColumnFile(
+                    file,
+                    new TestingKeyRetriever(Optional.of(KEY_FOOT), Optional.of(KEY_AGE), Optional.empty()),
+                    true,
+                    Optional.empty()))                          // NO prefix
+                    .isInstanceOfAny(ParquetCryptoException.class);
+        }
+    }
+
+    // ───────────────────────────────────  EXTRA CASES  ──────────────────────────────────
+
+    /**
+     * Two columns; footer encrypted; one column plaintext, other encrypted.
+     */
+    @Test
+    void mixedEncryptedAndPlaintextColumns()
+            throws IOException
+    {
+        File file = Files.createTempFile("pme‑mixed‑cols", ".parquet").toFile();
+        file.deleteOnExit();
+
+        // age column encrypted, id column plaintext
+        writeTwoColumnFile(file, /*encryptAge*/ true, /*encryptId*/ false);
+
+        // Provide keys for both footer & encrypted age column
+        Map<String, List<Integer>> data = readTwoColumnFile(file, new TestingKeyRetriever(Optional.of(KEY_FOOT), Optional.of(KEY_AGE), Optional.empty()));
+
+        verifySequence(data.get("age"), 100);
```

Comment:
same comment (we also need to verify that an encrypted file cannot be read without encryption keys. e.g. a test where the column key is not provided to the reader, resulting in an exception.)